### PR TITLE
Implement 64-bit rotation lowering for wasm2asm

### PR DESCRIPTION
**Pull request by alexcrichton**
Not much fancy here, but rather each operation is naively lowered inline to the
if/else chain to execute it.

---

FWIW I ended up finishing this implementation strategy before noticing the [preexising shim creation function](https://github.com/WebAssembly/binaryen/blob/51e515748e56c3ace6b3181e9ca8bf464c86f0e0/src/wasm2asm.h#L320-L374). Although poking around at that it didn't look immediately amenable to generating 64-bit arguments (due to the restriction that 64-bit can't be used I think?). If that's a better strategy though I can look into taking that path instead of doing the lowering inline like this!

### CODE REVIEW

**Comment by kripken** [comment]
Thanks!

The existing rotate generator looks 32-bit specific, yeah - even though the assert at the top suggests it was intended to eventually do more. I don't have a strong opinion on extending that vs adding a new function for the 64-bit case (which seems sufficiently different to make a single function difficult). But the approach of generating helper functions makes sense, I think, if the inline expansion is large as it is here. (The optimizer can inline if it makes sense later on.)

Another thing to consider here is that instead of adding a new test, we can enable an existing spec test. I just merged in the helper PR to clean up wasm2asm testing, so this should be simpler now. Looks like the spec test with rotates is i64.wast, though, which contains a million other things. So maybe that doesn't make sense yet. However, you will need to merge master and have the test & output in `/test/wasm2asm` now.

**Comment by alexcrichton** [comment]
Ok thanks for reviewing! I've moved the inline codegen to a separate function, generated only if necessary as part of the i64->i32 lowering pass. Other than that I hope to enable the `i64.wast` test soon for wasm2asm and get some more coverage!

File: src/passes/I64ToI32Lowering.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@
 #include "wasm-builder.h"
 #include "ir/names.h"
 
+#define ROTL64_NAME "__wasm_rotl_i64"
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: these appear to already be defined in `src/asmjs/shared-constants.h` which we could include here

File: src/passes/I64ToI32Lowering.cpp [hunk 2]
```diff
@@ -119,6 +121,147 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
     highBits->mutable_ = true;
     module->addGlobal(highBits);
     PostWalker<I64ToI32Lowering>::doWalkModule(module);
+
+    if (needRotl64)
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: `{, }` if the if body is one a separate line, please

File: src/passes/I64ToI32Lowering.cpp [hunk 3]
```diff
@@ -1475,6 +1664,8 @@ struct I64ToI32Lowering : public WalkerPass<PostWalker<I64ToI32Lowering>> {
   std::unordered_map<Name, TempVar> labelHighBitVars;
   std::vector<Index> freeTemps;
   Index nextTemp;
+  bool needRotl64;
```
Comment (kripken) [inline src/passes/I64ToI32Lowering.cpp]: need to ` = false` these

**Pull request merged**